### ISSUE TITLE
Segmentation fault in EVP_DigestSign() using SM2 private key

### ISSUE THREAD

**Issue by rwfranks** [openssl/openssl#23858]
#23816 the sequel

```
#include <stdio.h>
#include <openssl/core_names.h>
#include <openssl/param_build.h>
#include <openssl/bn.h>
#include <openssl/evp.h>
#define checkerr(x) if (!(x)) printf("error in " #x "\n")

int main()
{
        const char *privkey = "C3E0233E8EB9D0803C0D6784AB942A6762D6629B82FE878469868613566F2B20";
        BIGNUM *privbn = NULL;
        OSSL_LIB_CTX *libctx = NULL;
        OSSL_PARAM *params;
        OSSL_PARAM_BLD *bld = OSSL_PARAM_BLD_new();
        EVP_PKEY_CTX *ctx;
        EVP_PKEY *pkey = NULL;

        EVP_MD_CTX   *mctx;
        EVP_PKEY_CTX *pctx;
        const EVP_MD *md = EVP_sm3();
        unsigned char buffer[65];
        size_t buflen = sizeof(buffer);
        char *msgbuf = "arbitrary";
        size_t msglen = sizeof(msgbuf);
        checkerr( BN_hex2bn(&privbn, privkey) );
        checkerr( OSSL_PARAM_BLD_push_utf8_string( bld, OSSL_PKEY_PARAM_GROUP_NAME, "SM2", 0 ) );
        checkerr( OSSL_PARAM_BLD_push_BN( bld, OSSL_PKEY_PARAM_PRIV_KEY, privbn ) );
        params = OSSL_PARAM_BLD_to_param(bld);
        ctx = EVP_PKEY_CTX_new_from_name( libctx, "SM2", NULL );
        checkerr( EVP_PKEY_fromdata_init(ctx) );
        checkerr( EVP_PKEY_fromdata( ctx, &pkey, EVP_PKEY_KEYPAIR, params ) );
        BN_clear_free(privbn);
        EVP_PKEY_CTX_free(ctx);        OSSL_PARAM_free(params);
        OSSL_PARAM_BLD_free(bld);
        if ( pkey != NULL ) printf("so far, so good\n");

        mctx = EVP_MD_CTX_new();
        pctx = EVP_PKEY_CTX_new(pkey, NULL);
        EVP_PKEY_CTX_set1_id(pctx, "SM2_USERID", 0);
        EVP_MD_CTX_set_pkey_ctx(mctx, pctx);
        checkerr( EVP_DigestSignInit( mctx, NULL, md, NULL, pkey ) );
        checkerr( EVP_DigestSign( mctx, buffer, &buflen, msgbuf, msglen ) );    /* SEGV! */
        printf("signed\n");

        EVP_MD_CTX_free(mctx);
        EVP_PKEY_CTX_free(pctx);
        EVP_PKEY_free(pkey);
        printf("exit\n");
        return 0;
}
```

**Comment by nhorman** [comment]
I believe the SM2 algorithm requires the presence.of.the public portion of the key in the computation of a signature.  I didn't believe that myself at first, but i checked.and it's codified in the (admittedly) hard-to-parse translation of the specification, and it's codified in the digest computation in the openssl library

To fix it you need.to add the public key to your import data params

**Comment by t8m** [comment]
Nevertheless we should not crash during the signature operation. I do not think we can reject the import in fromdata function, however we should return a clear error if the public key is not set during the signature operation.

**Comment by nhorman** [comment]
Just add a check for the public key in ossl_sm2_compute_z_digest and raise an error if its not present then?

**Comment by mattcaswell** [comment]
This was fixed in master/3.3/3.2 by #23887.
It will be fixed in 3.1/3.0 by #24078 when that merges.
